gms/gossiper.cc

```diff
@@ -932,42 +979,8 @@ void gossiper::run() {
                 do_status_check().get();
             }
 
-            //
-            // Gossiper task runs only on CPU0:
-            //
-            //    - If endpoint_state_map or _live_endpoints have changed - duplicate
-            //      them across all other shards.
-            //    - Reschedule the gossiper only after execution on all nodes is done.
-            //
-            bool live_endpoint_changed = (_live_endpoints != _shadow_live_endpoints);
-            bool unreachable_endpoint_changed = (_unreachable_endpoints != _shadow_unreachable_endpoints);
-
-            if (live_endpoint_changed || unreachable_endpoint_changed) {
-                if (live_endpoint_changed) {
-                    _shadow_live_endpoints = _live_endpoints;
-                }
-
-                if (unreachable_endpoint_changed) {
-                    _shadow_unreachable_endpoints = _unreachable_endpoints;
-                }
-
-                container().invoke_on_all([this, live_endpoint_changed, unreachable_endpoint_changed, es = _endpoint_state_map] (gossiper& local_gossiper) {
-                    // Don't copy gossiper(CPU0) maps into themselves!
-                    if (this_shard_id() != 0) {
-                        if (live_endpoint_changed) {
-                            local_gossiper._live_endpoints = _shadow_live_endpoints;
-                        }
-
-                        if (unreachable_endpoint_changed) {
-                            local_gossiper._unreachable_endpoints = _shadow_unreachable_endpoints;
-                        }
-
-                        for (auto&& e : es) {
-                            local_gossiper._endpoint_state_map[e.first].set_alive(e.second.is_alive());
-                        }
-                    }
-                }).get();
-            }
+            // FIXME: _unreachable_endpoints should also be updated
+            duplicate_live_endpoints_on_change();
```

Comment:
This is a neat solution, since it avoids going into the rabbit hole of potentially breaking the semantics of the gossiper. We essentially query the latest state from shard 0 when needed.

But I think the original idea was that we change the shard 0 code to push the latest state to other shards as soon as it changes, but not once a second. 

What was the issue with that approach? Did you try it?

@asias what approach do you prefer? The pull approach, when get_live_members() fetches the latest stase, or the push approach, when the latest state is propagated to all of the shards on change?

Reply:
The first commit is just the method, alone is it's the original copy in `gossiper::run()` loop.

The second commit does the same copy logic as the loop does, but when the API is queried.

Note, the failure detector runs in *all shards*, so the updates goes from shardX to shard0 to other shards. So I don't think it's a good idea to copy all the time. I prefer to just copy on API query, which should be much less frequent.